shapeshift/web
# feat: gas oracle

## Description

- use unchained gas oracle for slow/average/fast estimation on both eip1559 and legacy fees
- lift approval/fee/build/broadcast functionality to common `utils/evm` which can be shared by defi opportunities and investor packages next (same tx building logic throughout)
- refactor existing swappers to utilize common evm utils, handle less than min quotes better, and general cleanup

Changes:
- all routes should show `insufficient sell amount` with estimated gas and receive value of 0 now (button shows min amount required)
- cowswap will not default to min amount when sell amount is too low to better display the right ratio diffs

NOTE:
unchained will need to be deployed with new gas oracle. this will be a blocking deploy on the backend to ensure enough time for testing and validating the gas oracle works as expected and continues to return accurate fees. once this is confirmed, we can deploy unchained to prod and release this feature as well (testing will be done in yeet)

## Pull Request Type


## Issue (if applicable)

## Risk

Med/High

## Testing

- Full regression test of all evm swappers
  - ensure 0 value quotes display estimated gas where applicable
  - ensure less than min quotes display gas and exchange rate, but show `Insufficient sell amount` and `Minimum` as expected
  - ensure expected swappers still show for the various trade pairs
  - ensure gas estimations on trade confirm are accurate and very close to confirmed tx fee
  - ensure we can still trade and don't hit any errors through full trade flow

### Engineering

:point_up: 

### Operations

:point_up: 

## Screenshots (if applicable)

### Thorchain
Current             |  New
:-------------------------:|:-------------------------:
![image](https://github.com/shapeshift/web/assets/35275952/0b1ca4bd-71b6-44bc-88d1-8179a63d7bf4) | ![image](https://github.com/shapeshift/web/assets/35275952/d84f43cd-9a84-4100-80f2-3c0ee437ea19)
![image](https://github.com/shapeshift/web/assets/35275952/0e5e5a63-84b1-4113-8596-b24ee90fac36) | ![image](https://github.com/shapeshift/web/assets/35275952/ffd2227e-5589-4a87-8ee1-7238e625b2f6)
![image](https://github.com/shapeshift/web/assets/35275952/d16fcb3f-fcce-4494-a44c-41f1fff57fa3) | ![image](https://github.com/shapeshift/web/assets/35275952/a75551db-0eaf-4366-8046-6ee442c2dff5)
![image](https://github.com/shapeshift/web/assets/35275952/e61f2716-dcf7-464d-88fe-cbc4b16c2ed0) | ![image](https://github.com/shapeshift/web/assets/35275952/bee78d01-87c0-47d8-b5e8-04ad8dc0cf1c)

![image](https://github.com/shapeshift/web/assets/35275952/bd21ca12-1dd8-45b0-b7c4-a70a41cb0828)

### CowSwap
Current             |  New
:-------------------------:|:-------------------------:
![image](https://github.com/shapeshift/web/assets/35275952/5610fae0-076a-4d51-8022-864285766fbe) | ![image](https://github.com/shapeshift/web/assets/35275952/f3d64bb1-185a-4c29-8e3f-b37cc3d6601c)
![image](https://github.com/shapeshift/web/assets/35275952/68393931-9aac-43d5-bea8-90c5593904f4) | ![image](https://github.com/shapeshift/web/assets/35275952/fed87df8-8e92-4810-a157-c4dea82e9283)
![image](https://github.com/shapeshift/web/assets/35275952/2f81a51a-5a3a-4f89-be07-21f6e9b185c2) | ![image](https://github.com/shapeshift/web/assets/35275952/979fff2a-0757-4676-bede-da59471b2690)
![image](https://github.com/shapeshift/web/assets/35275952/ec0a7bcd-082c-4466-9e45-bf87999c1a70) | ![image](https://github.com/shapeshift/web/assets/35275952/b28e2fdb-238e-44aa-be39-25f9c565baf2)

![image](https://github.com/shapeshift/web/assets/35275952/7d11e921-a497-47d9-ab2f-1f5072c39808)

### Zrx
Current             |  New
:-------------------------:|:-------------------------:
![image](https://github.com/shapeshift/web/assets/35275952/949ccb34-10be-4c66-9df8-909ebba142ff) | ![image](https://github.com/shapeshift/web/assets/35275952/5cb685af-e3d2-45a1-a477-0863a393a109)
![image](https://github.com/shapeshift/web/assets/35275952/e62b16f9-4cc6-41d8-bd5e-419baadc5cd1) | ![image](https://github.com/shapeshift/web/assets/35275952/ed758d9f-2889-48a9-a6f3-ec34aa64eb15)
![image](https://github.com/shapeshift/web/assets/35275952/1d56ab74-058a-42dc-a73b-28d375332a18) | ![image](https://github.com/shapeshift/web/assets/35275952/bf9c5c58-e65d-45c7-9cad-629e77237dfa)
![image](https://github.com/shapeshift/web/assets/35275952/4e080233-5528-463c-95d1-20e06253f105) | ![image](https://github.com/shapeshift/web/assets/35275952/ebc843b2-eb58-49d3-a7fa-4183b17c328e)

![image](https://github.com/shapeshift/web/assets/35275952/4a5b3a83-9762-47b9-b5f6-3d2499df3e5f)

### OneInch
Current             |  New
:-------------------------:|:-------------------------:
![image](https://github.com/shapeshift/web/assets/35275952/b2116ae5-3d5f-4e8c-8c4b-8d3cd229460f) | ![image](https://github.com/shapeshift/web/assets/35275952/12c47b53-84a8-4e5d-8628-a853bf02764a)
![image](https://github.com/shapeshift/web/assets/35275952/4801a702-7e2f-46e9-9532-d9e06b75fef0) | ![image](https://github.com/shapeshift/web/assets/35275952/2c4bcdb0-677e-4184-a1b3-6d026016e27b)
![image](https://github.com/shapeshift/web/assets/35275952/ec53b031-d179-4979-b8f3-5e8ab60a5bc3) | ![image](https://github.com/shapeshift/web/assets/35275952/8845e8db-9065-4857-a7dd-709efa1a6462)
![image](https://github.com/shapeshift/web/assets/35275952/76a9a16d-5367-4cb2-b068-34b979c3c912) | ![image](https://github.com/shapeshift/web/assets/35275952/fba5a3db-0c3b-4e36-91f4-443803ef57a8)

![image](https://github.com/shapeshift/web/assets/35275952/c5ae5d13-0816-43c0-9960-db4d53dc7bf5)
